server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/PartitionedClusterFactory.java

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.cluster.routing.allocation.allocator;
+
+/**
+ * A partitioned cluster factory must be able to divide all shards and nodes into mutually
+ * disjoint partitions. Allocation balancing will then be conducted sequentially for each partition.
+ * <p>
+ * If you can't partition your shards and nodes in this way, use
+ * {@link org.elasticsearch.cluster.routing.allocation.allocator.GlobalPartitionedClusterFactory}
+ */
+public interface PartitionedClusterFactory {
```

Comment:
I am not sure I understand the role of the factory? Looks like the underlying `PartitionedCluster` might as well be used directly - it is sort of a factory for the sorter  and weight funtion(s).

Reply:
Yeah the moment the PartitionedCluster is created is when we store the current set of weights. We could avoid the factory if we're happy for the weights to change under it.

Reply:
Yes, but we take a copy of them at the moment the `PartitionedCluster` instance is created. i.e. if someone updates the settings in the middle of a balance, the running balance will continue to use the copy it took when the balancing began. This is analogous to how the old singular `WeightFunction` was constructed in the call to `allocate()`